runtime/drivers/clickhouse/clickhouse.go

```diff
@@ -0,0 +1,180 @@
+package clickhouse
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/rilldata/rill/runtime/drivers"
+	"github.com/rilldata/rill/runtime/pkg/activity"
+	"github.com/rilldata/rill/runtime/pkg/priorityqueue"
+	"go.uber.org/zap"
+	"golang.org/x/sync/semaphore"
+
+	// import clickhouse driver
+	_ "github.com/ClickHouse/clickhouse-go/v2"
+)
+
+func init() {
+	drivers.Register("clickhouse", driver{})
+}
+
+type driver struct{}
+
+var maxOpenConnections = 20
+
+// Open connects to Clickhouse using std API.
+// Connection string format : https://github.com/ClickHouse/clickhouse-go?tab=readme-ov-file#dsn
+func (d driver) Open(config map[string]any, shared bool, client activity.Client, logger *zap.Logger) (drivers.Handle, error) {
+	if shared {
+		return nil, fmt.Errorf("clickhouse driver can't be shared")
+	}
+	dsn, ok := config["dsn"].(string)
+	if !ok {
+		return nil, fmt.Errorf("require dsn to open clickhouse connection")
+	}
+
+	db, err := sqlx.Open("clickhouse", dsn)
+	if err != nil {
+		return nil, err
+	}
+
+	// very roughly approximating num queries required for a typical page load
+	// TODO: copied from druid reevaluate
+	db.SetMaxOpenConns(maxOpenConnections)
+
+	err = db.Ping()
+	if err != nil {
+		return nil, fmt.Errorf("connection: %w", err)
+	}
+
+	conn := &connection{
+		db:      db,
+		config:  config,
+		logger:  logger,
+		metaSem: semaphore.NewWeighted(1),
+		olapSem: priorityqueue.NewSemaphore(maxOpenConnections - 1),
+	}
+	return conn, nil
+}
+
+func (d driver) Drop(config map[string]any, logger *zap.Logger) error {
+	return drivers.ErrDropNotSupported
+}
+
+func (d driver) Spec() drivers.Spec {
+	return drivers.Spec{}
+}
+
+func (d driver) HasAnonymousSourceAccess(ctx context.Context, src map[string]any, logger *zap.Logger) (bool, error) {
+	return false, fmt.Errorf("not implemented")
+}
+
+func (d driver) TertiarySourceConnectors(ctx context.Context, src map[string]any, logger *zap.Logger) ([]string, error) {
+	return nil, fmt.Errorf("not implemented")
+}
+
+type connection struct {
+	db       *sqlx.DB
+	config   map[string]any
+	logger   *zap.Logger
+	activity activity.Client
+
+	// logic around this copied from duckDB driver
+	// This driver may issue both OLAP and "meta" queries (like catalog info) against DuckDB.
+	// Meta queries are usually fast, but OLAP queries may take a long time. To enable predictable parallel performance,
+	// we gate queries with semaphores that limits the number of concurrent queries of each type.
+	// The metaSem allows 1 query at a time and the olapSem allows cfg.PoolSize-1 queries at a time.
+	// When cfg.PoolSize is 1, we set olapSem to still allow 1 query at a time.
+	// This creates contention for the same connection in database/sql's pool, but its locks will handle that.
+	metaSem *semaphore.Weighted
+	olapSem *priorityqueue.Semaphore
```

Comment:
Just to be clear Druid does not have any priority queue for queries. There are two types of isolation it provides - 
1. Hard isolation - We can create separate set of broker and historical tier and depending on any query property or context key for example priority we can route queries to a tier. Thus can physically have different resources to server different type of queries. Within a particular tier queries have FIFO order irrespective of priority so a higher priority query cannot preempt a lower one.
2. Soft isolation - We can limit max number of certain type of queries to run concurrently at a time using certain query property, for example a context key or interval its going to scan etc. This concept is calling laning, within a particular lane queries have FIFO order.

Reply:
I couldn't find anything that matches what druid allows or what this logic allows. 
There is a setting https://clickhouse.com/docs/en/operations/settings/settings#setting-os-thread-priority but I believe we need some mapping between current priority settings to this settings. Will evaluate more on this.

Reply:
As I understand we can use clickhouse's `os_thread_priority` setting for query prioritisation but would need following changes : 
1. We will have to write a translation from current priority numbers(0 to 100 with 100 being highest) to (-19 to 20 with 20 being least). 
2. Currently our queue ensures that if more queries than `maxOpenConnections` are submitted concurrently than higher priority queries are submitted first. This will not be possible and we will have to evaluate and increase `maxOpenConnections` so that prioritisation happens within clickhouse (default value in the driver is 10).

Please let me know if we want to make these changes now or can be taken in a separate PR.

Regarding concurrent data ingestion, yes clickhouse has a config for max concurrent insert and select queries but its a server property and can not be changed without restarting the server so will not be under our control for external clickhouse.

Reply:
Okay, let's not worry about `os_thread_priority` and other config at this stage. My thoughts were that the semaphores are something we added out of necessity for DuckDB and I wasn't sure if it would be needed for Clickhouse (like we don't have them for Druid). But I also understand keeping them there now that we anyway have them and if they seem to provide advantages. You decide :)

Reply:
Sure. I will add them as of now since they are already added. In future we can remove once we analyse clickhouse more.